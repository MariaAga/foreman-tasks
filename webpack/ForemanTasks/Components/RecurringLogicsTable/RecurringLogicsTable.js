import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from 'patternfly-react';
import { Table } from 'foremanReact/components/common/table';
import { STATUS } from 'foremanReact/constants';
import MessageBox from 'foremanReact/components/common/MessageBox';
import { translate as __ } from 'foremanReact/common/I18n';
import createTasksTableSchema from './RecurringLogicsSchema';

const RecurringLogicsTable = ({ status, recurringLogics, error }) => {
  if (status === STATUS.ERROR) {
    return (
      <MessageBox
        key="tasks-table-error"
        icontype="error-circle-o"
        msg={__(`Could not receive data: ${error}`)}
      />
    );
  }

  if (status === STATUS.PENDING && recurringLogics.length === 0) {
    return <Spinner size="lg" loading />;
  }

  if (recurringLogics.length === 0) {
    return <span>{__('No Recurring Logics')}</span>;
  }

  return (
    <div className="tasks-table">
      <Table
        key="tasks-table"
        columns={createTasksTableSchema()}
        rows={recurringLogics}
      />
    </div>
  );
};

RecurringLogicsTable.propTypes = {
  status: PropTypes.string,
  recurringLogics: PropTypes.array,
  error: PropTypes.string,
};

RecurringLogicsTable.defaultProps = {
  status: '',
  recurringLogics: [],
  error: null,
};
export default RecurringLogicsTable;
