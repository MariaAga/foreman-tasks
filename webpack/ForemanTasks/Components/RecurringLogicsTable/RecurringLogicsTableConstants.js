export const RECURRING_LOGICS = 'RECURRING_LOGICS';
