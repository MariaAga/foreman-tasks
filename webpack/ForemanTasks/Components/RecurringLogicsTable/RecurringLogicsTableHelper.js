export const getLastOccurrence = tasks => {
  const filteredTasks = [...tasks].filter(task => !!task.started_at);
  if (!filteredTasks.length) return '-';
  const lastTask = filteredTasks.sort((a, b) =>
    a.started_at < b.started_at ? 1 : -1
  )[0];
  return lastTask.started_at || '-';
};

export const getNextOccurrence = (state, tasks) => {
  const defaultLastOccurrence = '-';
  if (['cancelled', 'finished', 'disabled'].includes(state) || !tasks.length)
    return defaultLastOccurrence;
  const tasksList = [...tasks];
  const lastTask = tasksList.sort((a, b) =>
    a.start_at < b.start_at ? 1 : -1
  )[0];
  return lastTask.start_at;
};
