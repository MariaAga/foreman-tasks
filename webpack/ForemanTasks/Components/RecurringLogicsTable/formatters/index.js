export { taskCountCellFormatter } from './taskCountCellFormatter';
export { stateFormatter } from './stateFormatter';
export { actionCellFormatter } from './actionCellFormatter';
