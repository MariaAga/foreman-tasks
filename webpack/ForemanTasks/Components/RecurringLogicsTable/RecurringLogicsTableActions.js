import { getTableItemsAction } from 'foremanReact/components/common/table';
import { RECURRING_LOGICS } from './RecurringLogicsTableConstants';

export const getTableItems = () =>
  getTableItemsAction(
    RECURRING_LOGICS,
    {},
    `/foreman_tasks/api/recurring_logics${window.location.search}`
  );
