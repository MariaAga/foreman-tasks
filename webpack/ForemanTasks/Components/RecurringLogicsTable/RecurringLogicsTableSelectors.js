import { createSelector } from 'reselect';
import { selectAPIResponse } from 'foremanReact/redux/API/APISelectors';

import {
  getLastOccurrence,
  getNextOccurrence,
} from './RecurringLogicsTableHelper';
import { RECURRING_LOGICS } from './RecurringLogicsTableConstants';

const selectResponse = state => selectAPIResponse(state, RECURRING_LOGICS);

export const selectResults = createSelector(selectResponse, ({ results }) =>
  results
    ? results.map(result => ({
        cronLine: result.cron_line,
        taskCount: result.tasks.length,
        action: result.tasks[0] ? result.tasks[0].action : '-',
        lastOccurrence: getLastOccurrence(result.tasks),
        nextOccurrence: getNextOccurrence(result.state, result.tasks),
        currentIteration: result.iteration,
        iterationLimit: result.max_iteration,
        repeatUntil: result.end_time,
        state: result.state,
        id: result.id,
        canEdit: result.can_edit,
        TaskGroupId: result.task_group_id,
      }))
    : []
);
