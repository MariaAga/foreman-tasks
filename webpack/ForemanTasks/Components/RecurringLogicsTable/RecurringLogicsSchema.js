import React from 'react';
import {
  column,
  headerFormatterWithProps,
  cellFormatter,
} from 'foremanReact/components/common/table';
import { translate as __ } from 'foremanReact/common/I18n';
import { actionNameCellFormatter } from '../TasksTable/formatters';
import {
  taskCountCellFormatter,
  stateFormatter,
  actionCellFormatter,
} from './formatters';

const headFormat = [headerFormatterWithProps];
const cellFormat = [cellFormatter];

export const limitCellFormatter = limit =>
  cellFormatter(limit || <i>{__('Unlimited')}</i>);

const createTasksTableSchema = () => [
  column('cronLine', __('Cron line'), headFormat, [
    actionNameCellFormatter('foreman_tasks/recurring_logics'),
  ]),
  column('taskCount', __('Task count'), headFormat, [taskCountCellFormatter]),
  column('action', __('Action'), headFormat, cellFormat),
  column('lastOccurrence', __('Last occurrence'), headFormat, cellFormat),
  column('nextOccurrence', __('Next occurrence'), headFormat, cellFormat),
  column('currentIteration', __('Current iteration'), headFormat, cellFormat),
  column('iterationLimit', __('Iteration limit'), headFormat, [
    limitCellFormatter,
  ]),
  column('repeatUntil', __('Repeat until'), headFormat, [limitCellFormatter]),
  column('state', __('State'), headFormat, [stateFormatter]),
  column('canEdit', __(' '), headFormat, [actionCellFormatter()]),
];

export default createTasksTableSchema;
