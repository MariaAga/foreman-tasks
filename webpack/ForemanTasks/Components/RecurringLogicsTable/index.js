import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  selectAPIErrorMessage,
  selectAPIStatus,
} from 'foremanReact/redux/API/APISelectors';
import { selectResults } from './RecurringLogicsTableSelectors';

import { RECURRING_LOGICS } from './RecurringLogicsTableConstants';
import RecurringLogicsTable from './RecurringLogicsTable';
import { getTableItems } from './RecurringLogicsTableActions';

const ConnectedRecurringLogicsTable = () => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getTableItems());
  }, [dispatch]);
  const recurringLogics = useSelector(selectResults);
  const error = useSelector(state =>
    selectAPIErrorMessage(state, RECURRING_LOGICS)
  );
  const status = useSelector(state => selectAPIStatus(state, RECURRING_LOGICS));

  return (
    <RecurringLogicsTable
      recurringLogics={recurringLogics}
      error={error}
      status={status}
    />
  );
};

export default ConnectedRecurringLogicsTable;
