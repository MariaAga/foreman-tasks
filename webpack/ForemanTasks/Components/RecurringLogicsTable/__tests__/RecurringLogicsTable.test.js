import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import { STATUS } from 'foremanReact/constants';
import RecurringLogicsTable from '../RecurringLogicsTable';

jest.mock('../RecurringLogicsSchema');

const recurringLogics = [{ logic: 1 }];
const fixtures = {
  'render with no Props': {},
  'render with no recurringLogics': {
    recurringLogics: [],
    status: STATUS.RESOLVED,
  },
  'render with error Props': {
    error: 'some error message',
    status: STATUS.ERROR,
  },
  'render pending': {
    status: STATUS.PENDING,
  },
  'render with recurringLogics': {
    recurringLogics,
    status: STATUS.RESOLVED,
  },
};

describe('RecurringLogicsTable', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(RecurringLogicsTable, fixtures));
});
