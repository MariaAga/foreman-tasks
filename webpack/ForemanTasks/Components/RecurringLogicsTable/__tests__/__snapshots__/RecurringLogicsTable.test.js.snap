// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecurringLogicsTable rendering render pending 1`] = `
<Spinner
  className=""
  inline={false}
  inverse={false}
  loading={true}
  size="lg"
/>
`;

exports[`RecurringLogicsTable rendering render with error Props 1`] = `
<MessageBox
  icontype="error-circle-o"
  key="tasks-table-error"
  msg="Could not receive data: some error message"
/>
`;

exports[`RecurringLogicsTable rendering render with no Props 1`] = `
<span>
  No Recurring Logics
</span>
`;

exports[`RecurringLogicsTable rendering render with no recurringLogics 1`] = `
<span>
  No Recurring Logics
</span>
`;

exports[`RecurringLogicsTable rendering render with recurringLogics 1`] = `
<div
  className="tasks-table"
>
  <Table
    key="tasks-table"
    rows={
      Array [
        Object {
          "logic": 1,
        },
      ]
    }
  />
</div>
`;
